sale_coupon_advanced/tests/test_sale_coupon_forced_reward_product.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo.addons.sale_coupon.tests.common import TestSaleCouponCommon
+
+
+class TestSaleCouponForcedRewardProduct(TestSaleCouponCommon):
+    def setUp(self):
+        super().setUp()
+
+        self.env["sale.coupon.program"].search([]).write({"active": False})
+
+        # create partner for sale order.
+        self.john_doe = self.env["res.partner"].create(
+            {"name": "John Doe", "email": "john.doe@example.com", "lang": "en_US"}
+        )
+
+        self.order_forced = self.env["sale.order"].create(
+            {
+                "partner_id": self.john_doe.id,
+                "order_line": [
+                    (
+                        0,
+                        False,
+                        {
+                            "product_id": self.product_A.id,
+                            "name": "1 Product A",
+                            "product_uom": self.uom_unit.id,
+                            "product_uom_qty": 2.0,
+                        },
+                    ),
+                ],
+            }
+        )
+
+        self.program_forced = self.env["sale.coupon.program"].create(
+            {
+                "name": "Buy anything, B are free",
+                "promo_code_usage": "no_code_needed",
+                "reward_type": "product",
+                "reward_product_id": self.product_B.id,
+                "rule_products_domain": "[('sale_ok', '=', True)]",
+                "is_reward_product_forced": True,
+                "active": True,
+                "sequence": 1,
+            }
+        )
+        self.reward_product = self.program_forced.reward_product_id
+        self.discount_product = self.program_forced.discount_line_product_id
+
+    def tearDown(self):
+        super().tearDown()
+        self.order_forced.unlink()
```

Comment:
It should be already unlinked after tests are executed.

Reply:
Should or will ?

Reply:
Tests are running in a dedicated environment. The database is rollbacked at different stages depending on the test class you are inheriting.
You don't need to unlink everything by yourself.

Reply:
Fixed